Assets/MixedRealityToolkit.Services/InputSystem/Utilities/ScaleMeshEffect.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.Toolkit.Input.Utilities
+{
+    /// <summary>
+    /// On Unity UI components the unity_ObjectToWorld matrix is not the transformation matrix of the local 
+    /// transform the Graphic component lives on, but that of it's parent Canvas. Many MRTK/Standard shader 
+    /// effects require object scale to be known. To solve this issue the ScaleMeshEffect will store scaling 
+    /// information into UV channel attributes during UI mesh construction. Ideally we would store the scale 
+    /// in one attribute but UGUI only supports two scalers per attribute (even in the tangent attribute).
+    /// </summary>
+    [RequireComponent(typeof(RectTransform))]
+    public class ScaleMeshEffect : BaseMeshEffect
+    {
+        /// <summary>
+        /// Enforces the parent canvas uses UV channel attributes.
+        /// </summary>
+        protected override void Awake()
+        {
+            base.Awake();
+
+            // Make sure the parent canvas is configured to use UV channel attributes for scaling data.
+            var canvas = GetComponentInParent<Canvas>();
+
+            if (canvas != null)
+            {
+                canvas.additionalShaderChannels |= AdditionalCanvasShaderChannels.TexCoord2;
+                canvas.additionalShaderChannels |= AdditionalCanvasShaderChannels.TexCoord3;
+            }
+        }
+
+        /// <summary>
+        /// Stores scaling information into UV channel attributes during UI mesh construction.
+        /// </summary>
+        /// <param name="vh">The vertex helper to populate with new vertex data.</param>
+        public override void ModifyMesh(VertexHelper vh)
+        {
+            var rectTransform = transform as RectTransform;
+            var scale = new Vector2(rectTransform.rect.width * rectTransform.localScale.x,
```

Comment:
>localScale [](start = 77, length = 10)

Will this work for nested rectTransforms with non-1 localScale?

Ex:
Canvas
  > Parent (scale not 1.0)
    > ImageWithScaleMeshEffect

You might need to iterate through the hierarchy accumulating localScales.  Or alternately, use rectTransform.scale / canvas.scale maybe?

Reply:
That's a good point, it might not? Let me verify and get back to you.

Reply:
I just tested this on PressableButtonsTestCanvas by scaling the PressableButtons RectTransform. Luckily everything still works I believe due to the way vertices get rebuilt.